### ISSUE TITLE
GR00T outputs absolute action only?

### ISSUE THREAD

**Issue by jkim50104** [NVIDIA/Isaac-GR00T#490]
Hello, and thank you for sharing this great work!

I’ve trained my custom robot using this policy, and based on open-loop evaluation, the performance looks quite good so far. However, I have a question regarding the action representations during inference.

During training, I configured the left and right arms to use relative actions, while the remaining joints were trained with absolute actions. When running inference, though, it appears that the policy outputs absolute joint actions regardless of whether a joint was trained with a relative configuration. In other words, even for joints trained with the relative option, the actions returned internally seem to be absolute.

Is this the expected behavior? If so, could you clarify how relative vs. absolute configurations are handled during inference?

Thanks again for the excellent work, and I appreciate any clarification you can provide.

**Comment by HomerIsAFool** [comment]
I encountered the same problem, the output from the model is always absolute angles...
Also, i found that the gt action input in the training process is also absolute..

**Comment by xiaotongc0** [comment]
Hi, thanks for your interest to the work. I studied the dataflow of the current N1.6 code and seems you need to double check the two conditions to enable relative in training. And it's true the output will always be absolute. Below are details.

How Relative/Absolute Action Is Configured                            
                                                                        
There are two layers of configuration that control whether relative actions are used:
                                  
1. Global flag: use_relative_action in the model config (gr00t/configs/model/gr00t_n1d6.py:52), defaults to False.

2. Per-action-group representation: Each embodiment config in gr00t/configs/data/embodiment_configs.py specifies ActionRepresentation.RELATIVE or ABSOLUTE per joint group (e.g., left_arm → RELATIVE, hands → ABSOLUTE).
These two conditions are AND-ed together — a joint group is only treated as relative if both the global flag is True and the per-group config is RELATIVE. See state_action_processor.py:330: if action_config.rep == ActionRepresentation.RELATIVE and self.use_relative_action:

Potential Problem and Fix

The most likely cause is that use_relative_action was not set to True during training. If you wrote a custom training script (rather than using launch_finetune.py), you need to explicitly set:
config.model.use_relative_action = True

Our provided finetuning entrypoint (gr00t/experiment/launch_finetune.py:70) does this automatically. If you used it, this should already be handled — but it's worth double-checking your checkpoint's processor_config.json to confirm use_relative_action is true.

Also note: when use_relative_action=True, the dataset statistics must include a relative_action key with per-group stats. If these are missing, you'll get a ValueError at initialization.

Regarding inference output always being absolute — this is expected behavior. The pipeline converts absolute → relative during training, and converts back relative → absolute during inference. The final output is always absolute joint positions, which is what the robot controller expects.

Full Pipeline Reference

Training (forward):
- Raw dataset actions are in absolute joint space
- apply_action() checks each action group: if use_relative_action=True and action_config.rep=RELATIVE, it converts absolute → relative by subtracting the last state timestep as a reference frame
- Normalization is then applied using relative_action statistics (not action statistics) for those groups
- The diffusion model (ActionHead) is representation-agnostic — it simply learns whatever distribution it's fed

Inference (reverse):
- The model outputs normalized actions in the same representation space it was trained on
- unapply_action() first denormalizes using the corresponding statistics
- Then, if conditions match, it converts relative → absolute by adding back the last state timestep
- Final output is always absolute actions sent to the robot

Checkpoint persistence:
- use_relative_action is saved into processor_config.json during training
- from_pretrained() restores it automatically at inference time, so you don't need to set it again manually

**Comment by mohitydv09** [comment]
@xiaotongc0 Thanks for the detailed breakdown. One thing I’m curious about is: in the UMI paper, where they introduced relative action prediction, an important aspect is training the model without using the current absolute state of the end-effector (EEF). This helps avoid issues with data distribution shifts when the data is collected on a different robot, since the model only learns to predict deltas from the image.

However, it seems this cannot be done with the current Gr00t codebase, as the model requires the current state of the robot as input. I’m imagining training the robot using a dataset collected with UMI, where the absolute EEF from UMI may have a distribution shift compared to that of the robot. Is there a way to achieve this?
